intermediate_source/TP_tutorial.rst

```diff
@@ -0,0 +1,282 @@
+
+.. figure:: /_static/img/distributed/loss_parallel.png
+   :width: 100%
+   :align: center
+   :alt: loss parallel
+
+   Figure 2. Cross-entropy loss forward computation with loss parallel on one GPU. Blue represents sharded tensors; green represents replicated tensors; yellow represents tensors with partial values (to be all-reduced). Black arrows are local computations; red arrows are functional collectives among GPUs.
+
+In the PyTorch Tensor Parallel API, Loss Parallel can be enabled via a context manager ``loss_parallel``, with which one can directly use ``torch.nn.functional.cross_entropy`` or ``torch.nn.CrossEntropyLoss`` without modifying other parts of their code.
+
+To apply Loss Parallel, the model predictions, usually of the shape ``[batch size, sequence length, vocabulary size]``, should be sharded on the vocabulary dimension. This can be easily done via marking the output layouts of the last linear projection layer output:
+
+.. code-block:: python
+    model = parallelize_module(
+        model,
+        tp_mesh,
+        {
+            "output": ColwiseParallel(
+                input_layouts=Shard(1),
+                use_local_output=False,
+            ),
+            "norm": SequenceParallel(),
+            "layers.0": PrepareModuleInput(
+                input_layouts=(Replicate(),),
+                desired_input_layouts=(Shard(1),),
+            ),
+        },
+    )
+
+In the code above, we also apply Sequence Parallel to the norm layer before output. We apply ``use_local_output=False`` to let the output stay as a DTensor, to work with the ``loss_parallel`` context manager. After that, one can simply call the cross_entropy loss function as is shown below. Note that the backward computation also needs to happen within the context.
+
+.. code-block:: python
+    import torch.nn.functional as F
+    from torch.distributed.tensor.parallel import loss_parallel
+
+    pred = model(input_ids)
+    with loss_parallel():
+        # assuming pred and labels are of the shape [batch, seq, vocab]
+        loss = F.cross_entropy(pred.flatten(0, 1), labels.flatten(0, 1))
+        loss.backward()
+
+
+How to combine Tensor Parallel with Fully Sharded Data Parallel together
+------------
+
+
+Now that we have shown how to apply Tensor/Sequence Parallel to the model, let us also take a look at how Tensor Parallel and Fully Sharded Data Parallel could work together. Since Tensor Parallelism incurs communications that’s blocking the computation, we want to make sure it runs within a fast communication channel (i.e. NVLink). In practice, we usually apply Tensor Parallel within each host, and apply Fully Sharded Data Parallel across the hosts.
```

Comment:
```suggestion
Now that we have shown how to apply Tensor/Sequence Parallel to the model, let us also take a look at how Tensor Parallel and Fully Sharded Data Parallel could work together. Since Tensor Parallelism incurs communications that’s blocking the computation, we want to make sure it runs within a fast communication channel, such as NVLink. In practice, we usually apply Tensor Parallel within each host, and apply Fully Sharded Data Parallel across the hosts.
```